### PR TITLE
Fix New-TestResources.ps1 to run pre/post deployment script

### PR DESCRIPTION
**Pull request by ckairen**
Fix bug in which New-TestResources.ps1 doesn't call test-resources-pre.ps1 and test-resources-post.ps1 script before and after template deployment from a bicep file.

### CODE REVIEW

File: eng/common/TestResources/New-TestResources.ps1
Comment (heaths) [inline eng/common/TestResources/New-TestResources.ps1:480]: IMO, this should be the `originalFilePath` since bicep would've failed to compile and broken the script earlier.

Comment (heaths) [reply eng/common/TestResources/New-TestResources.ps1:480]: Alternatively, you could do something like:

```powershell
$msg = if ($templateFile.jsonFilePath -ne $templateFile.originalFilePath) {
  "Deployment template '$($templateFile.jsonFilePath)' ($templateFile.originalFilePath) to resource group ..." } else {
  "deployment template '$($templateFile.jsonFilePath)' to resource group ..."
}
```

File: eng/common/TestResources/New-TestResources.ps1
Comment (heaths) [inline eng/common/TestResources/New-TestResources.ps1:585]: I don't remember from the original PR to add bicep support, but I think this should go into $env:TEMP and we don't really have to worry about it - not in the repo. No worries about it getting checked in then either.

**Review by heaths — APPROVED** [review]
Overall great change. Thank you! I do have a few suggestions.

File: eng/common/TestResources/New-TestResources.ps1
Comment (weshaggard) [reply eng/common/TestResources/New-TestResources.ps1:585]: It does generate to the temp directory but I'd like to try and make an attempt to clean it up if possible as I don't like an ever growing temp directory.

File: eng/common/TestResources/New-TestResources.ps1
Comment (weshaggard) [reply eng/common/TestResources/New-TestResources.ps1:480]: @heaths why do you think this is an error case where the bicep compile failed?

Comment (heaths) [reply eng/common/TestResources/New-TestResources.ps1:480]: If the bicep file has a syntax error, a listKeys* function is wrong (I did that once, and while the bicep file compiled the deployment error was misleading), etc. Plenty of reasons user input may fail to compile.

File: eng/common/TestResources/New-TestResources.ps1
Comment (heaths) [reply eng/common/TestResources/New-TestResources.ps1:585]: Fair enough, but you should turn out temporary %TEMP% file clean up in Windows. In Settings, search for "Delete temporary files".

File: eng/common/TestResources/New-TestResources.ps1
Comment (weshaggard) [reply eng/common/TestResources/New-TestResources.ps1:480]: I understand that the bicep build might fail but I'm not sure what this message has to do with that failure case.

Comment (heaths) [reply eng/common/TestResources/New-TestResources.ps1:480]: I never said anything about a failure case. I saw I think the message should point to the original file in some way. *IIF* the bicep compilation failed, we never would've gotten that far anyway so it doesn't matter.

Comment (weshaggard) [reply eng/common/TestResources/New-TestResources.ps1:480]: OK I think I misunderstood. I'm good either way.

**Review by weshaggard — APPROVED** [review]
The changes look good but as we talked lets wait until after releases next week before we get this merged, unless there is someone actively blocked by this.

File: eng/common/TestResources/New-TestResources.ps1
Comment (benbp) [inline eng/common/TestResources/New-TestResources.ps1:175]: Functionally, this is great. I wonder if there's an opportunity to do a little more refactoring here to simplify the downstream code and make it easier to avoid bugs in the future (e.g. using the wrong file path).

If you add the directory to the templateFile object here, you can avoid all the calls to `Split-Path` below and instead use `$preDeploymentScript = Join-Path $templateFile.Location 'test-resources-pre.ps1'`


```suggestion
            $templateFile = @{
                location = $_.DirectoryName
                originalFilePath = $_.FullName;
                jsonFilePath = $_.FullName
            }
            if ($_.Extension -eq '.bicep') {
                $templateFile.jsonFilePath = (BuildBicepFile $_)
            }
            $templateFiles += $templateFile
```

Just a suggestion though, feel free to ignore :)

**Review by benbp — APPROVED** [review]

File: eng/common/TestResources/New-TestResources.ps1
Comment (benbp) [inline eng/common/TestResources/New-TestResources.ps1:175]: Functionally, this is great. I wonder if there's an opportunity to do a little more refactoring here to simplify the downstream code and make it easier to avoid bugs in the future (e.g. using the wrong file path).

If you add the directory to the templateFile object here, you can avoid all the calls to `Split-Path` below and instead use `$preDeploymentScript = Join-Path $templateFile.Location 'test-resources-pre.ps1'`


```suggestion
            $templateFile = @{
                location = $_.DirectoryName
                originalFilePath = $_.FullName;
                jsonFilePath = $_.FullName
            }
            if ($_.Extension -eq '.bicep') {
                $templateFile.jsonFilePath = (BuildBicepFile $_)
            }
            $templateFiles += $templateFile
```

Just a suggestion though, feel free to ignore :)

**Comment by benbp** [comment]
Thanks for tracking this bug down!

File: eng/common/TestResources/New-TestResources.ps1
Comment (heaths) [inline eng/common/TestResources/New-TestResources.ps1]: I'm honestly not a fan of this change, @benbp. The temp file that gets returned and printed everywhere is an implementation detail of this script with no obvious tie-back to the original .bicep file. At the *very least* `BuildBicepFile` should print the source and destination paths, but that's still requiring users to look up through the logs to find it. How it was with the object I thought was a great idea, so you could even print both if someone needed to find the actual JSON template (I've ran into a few issues where this was needed with bicep files).

Comment (ckairen) [reply eng/common/TestResources/New-TestResources.ps1]: I think you might be commenting on an outdated commit? Sorry for the messy commits.

Comment (heaths) [reply eng/common/TestResources/New-TestResources.ps1]: No, at the time I commented, the variable was just a single string path. You had it as a multi-property object but Ben asked you to change it. Did you chances it back to an object? We shouldn't be printing the path to a temp JSON file that the user can't use to diagnose failures.

Comment (benbp) [reply eng/common/TestResources/New-TestResources.ps1]: I think it is an outdated diff. The main code is object based: https://github.com/Azure/azure-sdk-tools/blob/main/eng/common/TestResources/New-TestResources.ps1#L170-L171

I think it may actually be helpful to print both paths, because when debugging bicep scripts, I often need to see the compiled output to help diagnose the problem.

Comment (heaths) [reply eng/common/TestResources/New-TestResources.ps1]: Yes, that's what I originally commented as well. The commit that had the original and transformed names as properties was best. My only comment from then was that the message printed *when* the two were different should list both. From the number of people I help set up their test resources (rather, customize them) it's clear expecting them to know how it works e.g., that it transforms the bicep file into json behind the scenes would be unnecessary.

Comment (benbp) [reply eng/common/TestResources/New-TestResources.ps1]: Yep makes sense. I plan to roll in those changes with my fix to https://github.com/Azure/azure-sdk-tools/issues/1824 (I promise I'll get to it @pakrym, just dealing with weird windows encryption API issues).

**Pull request merged**
